gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/InboundMessageMergingService.java

```diff
@@ -0,0 +1,162 @@
+import uk.nhs.adaptors.pss.translator.model.EbxmlReference;
+import uk.nhs.adaptors.pss.translator.util.XmlParseUtilService;
+
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.ValidationException;
+import javax.xml.transform.TransformerException;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static uk.nhs.adaptors.connector.model.MigrationStatus.EHR_EXTRACT_TRANSLATED;
+import static uk.nhs.adaptors.pss.translator.model.NACKReason.EHR_EXTRACT_CANNOT_BE_PROCESSED;
+import static uk.nhs.adaptors.pss.translator.util.XmlUnmarshallUtil.unmarshallString;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class InboundMessageMergingService {
+
+    private final PatientAttachmentLogService patientAttachmentLogService;
+    private final AttachmentHandlerService attachmentHandlerService;
+    private final AttachmentReferenceUpdaterService attachmentReferenceUpdaterService;
+    private final MigrationStatusLogService migrationStatusLogService;
+    private final XmlParseUtilService xmlParseUtilService;
+    private final ObjectMapper objectMapper;
+    private final XPathService xPathService;
+    private final FhirParser fhirParser;
+    private final BundleMapperService bundleMapperService;
+    private final PatientMigrationRequestDao migrationRequestDao;
+    private final NackAckPreparationService nackAckPreparationService;
+    private static final String CONVERSATION_ID_HAS_NOT_BEEN_GIVEN = "Conversation Id has not been given";
+
+    public boolean canMergeCompleteBundle(String conversationId) throws ValidationException {
+
+        if (!StringUtils.hasText(conversationId)) {
+            throw new ValidationException(CONVERSATION_ID_HAS_NOT_BEEN_GIVEN);
+        }
+
+        var undeletedLogs = getUndeletedLogsForConversation(conversationId);
+        return undeletedLogs.stream().allMatch(log -> log.getUploaded().equals(true));
+    }
+
+    public void mergeAndBundleMessage(String conversationId) throws JAXBException, JsonProcessingException {
+
+        if (!StringUtils.hasText(conversationId)) {
+            throw new ValidationException(CONVERSATION_ID_HAS_NOT_BEEN_GIVEN);
+        }
+
+        PatientMigrationRequest migrationRequest = migrationRequestDao.getMigrationRequest(conversationId);
+        var inboundMessage = objectMapper.readValue(migrationRequest.getInboundMessage(), InboundMessage.class);
+        RCMRIN030000UK06Message payload = unmarshallString(inboundMessage.getPayload(), RCMRIN030000UK06Message.class);
+
+        try {
+
+            var attachmentLogs = getUndeletedLogsForConversation(conversationId);
+            var attachmentsContainSkeletonMessage = attachmentLogs.stream().anyMatch(log -> log.getSkeleton().equals(true));
+
+            if (attachmentsContainSkeletonMessage) {
```

Comment:
Maybe contents of this if statement could be moved to a separate method, as the whole try is quite large now